### PR TITLE
Report Unknown Doc-Comment Tags in Slice

### LINKED ISSUE #2088: @remarks results in bad generated Swift code
I added `@remarks` to a Slice operation and the Swift generated code no longer compiles.

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR fixes #2088.
Now, the Slice parser rejects unknown doc-comment tags (including `@remarks`), making it's situation impossible to hit.

It also adds a check which rejects multi-line `@see` tags, since these are disallowed and `@see` should always be of the form: `@see IDENTIFIER` and nothing else.

When the parser encounters either of these cases, we discard (ignore) the offending line, and issue a _warning_ for it.
I think that the predominant philosophy in compilers is that: "a badly formatted comment should never break your build".
Since, on a logical level, compilers _should be_ ignoring all comments in your source file.

Since we currently have no validation of comments, I also added a new "InvalidComment" diagnostic category, which this PR uses.

### CODE REVIEW

File: cpp/src/Slice/Parser.cpp [hunk 1]
```diff
@@ -777,95 +775,99 @@ Slice::Contained::parseComment(function<string(string, string)> linkFormatter, b
...
+    const string seeTag = "@see";
     const string returnTag = "@return";
     const string deprecatedTag = "@deprecated";
-    const string seeTag = "@see";
     for (; i != lines.end(); ++i)
     {
         const string l = IceInternal::trim(*i);
-        string line;
-        if (parseCommentLine(l, paramTag, true, name, line))
+        string lineText;
```
Comment (InsertCreativityHere) [inline cpp/src/Slice/Parser.cpp:797]: Renamed to avoid shadowing the `line` method which is defined on this class.

**Comment by InsertCreativityHere** [comment]
After thinking about it more. I think even a warning is too harsh.
We shouldn't issue any diagnostics at all for unknown tags.
Just silently ignore them, and move on.

> There is no sensible reason to tolerate them.

It's important to remember that there's **two** purposes to doc-comments.
One is that we use them to generate doc-comments in the mapped languages, yes.
But another, equally important one, is to document the Slice definitions themselves.
And Slice doc-comments support the _full_ breadth of doxygen.

It just so happens that the _Slice parser_ is only aware of a very limited subset of these.
Because supporting and mapping tags takes time and there's just better things to spend that time on.

But we do not want to restrict users to only using the very small subset of Doxygen tags that we map in the generated code.
This would rob them of the extensive functionality doxygen provides. And if there's any users out there who use this functionality (which we document as providing), the suggested change would completely break their builds.

**TL;DR**

Just because we the parser doesn't generate code for a doc-comment tag, does not make that tag invalid.
There is no such thing as "Slice comment tags" there are only "Doxygen comment tags",
and we should make no attempt to convert 'slice2xxx' into a Doxygen tag validator.

**Comment by InsertCreativityHere** [comment]
After discussion, we've come to the following middle ground:

1) We agree that we shouldn't just slap the unknown tags into the code, we _have_ to ignore and discard it.
2) We should issue warnings for unknown tags; silently ignoring them would be too surprising.
3) In the future, we should add parser support for other common Doxygen tags, to reduce the number of warnings users who are familiar with Doxygen will encounter.

File: cpp/test/Slice/errorDetection/WarningInvalidComment.err [hunk 2]
```diff
@@ -0,0 +1,3 @@
+WarningInvalidComment.ice:10: warning: ignoring unknown doc tag '@remarks' in comment
```
Comment (bernardnormier) [inline cpp/test/Slice/errorDetection/WarningInvalidComment.err:1]: Should we refer to these comments as "doc comment[s]"?

I find the diagnostic "ignored xxx in comment" odd as everything in plain comment should be ignored.
If I write:
```
// @foo in a plain comment
```

I don't anticipate any diagnostic.

File: cpp/src/Slice/Parser.h [hunk 3]
```diff
@@ -38,7 +38,8 @@ namespace Slice
     {
         All,
         Deprecated,
-        InvalidMetadata
+        InvalidMetadata,
+        InvalidComment
```
Comment (bernardnormier) [inline cpp/src/Slice/Parser.h:42]: Should be InvalidDocComment. A plain comment is always valid and ignored.

File: cpp/src/Slice/Parser.h [hunk 4]
```diff
@@ -252,9 +253,7 @@ namespace Slice
 
         /// Contains all introductory lines up to the first tag.
         StringList overview() const;
-        /// Contains unrecognized tags.
-        StringList misc() const;
-        /// Targets of @see tags.
+        /// Targets of '@see' tags.
```
Comment (bernardnormier) [inline cpp/src/Slice/Parser.h:256]: We should rename this class `DocComment`.

**Review by bernardnormier — APPROVED** [review]

**Review by externl — APPROVED** [review]

**Comment by InsertCreativityHere** [comment]
I agree with renaming `Comment` to doc comment, and have opened an issue to do it: #3141.
I'd rather delay all the renaming of `comment` -> `doc comment` until that PR, so it's all done together.
